src/Main.elm

```diff
@@ -1,16 +1,411 @@
+                Ok data ->
+                    data
+
+                -- @TODO: error reporting
+                Err err ->
+                    []
+    in
+        ( { audienceFolders = audienceFolder
+          , audiences = audiences
+          , currentPath = []
+          , selectedFolder = Nothing
+          }
+        , Cmd.none
+        )
+
+
+
+-- DECODERS
+
+
+audienceFoldersDecoder : String -> Result String (List AudienceFolder)
+audienceFoldersDecoder json =
+    json
+        |> Decode.decodeString (Decode.at [ "data" ] <| Decode.list audienceFolderDecodeItem)
+
+
+audienceFolderDecodeItem : Decoder AudienceFolder
+audienceFolderDecodeItem =
+    Decode.succeed AudienceFolder
+        |: (field "id" Decode.int)
+        |: (field "name" Decode.string)
+        |: (field "parent" <| Decode.maybe Decode.int)
+
+
+audiencesDecoder : String -> Result String (List Audience)
+audiencesDecoder json =
+    json
+        |> Decode.decodeString (Decode.at [ "data" ] <| Decode.list audiencesDecoderItem)
+
+
+audiencesDecoderItem : Decoder Audience
+audiencesDecoderItem =
+    Decode.succeed Audience
+        |: (field "id" Decode.int)
+        |: (field "name" Decode.string)
+        |: (field "type"
+                Decode.string
+                |> Decode.andThen audienceTypeDecoder
+           )
+        |: (field "folder" <| Decode.maybe Decode.int)
+
+
+audienceTypeDecoder : String -> Decoder AudienceType
+audienceTypeDecoder audienceType =
+    case audienceType of
+        "user" ->
+            Decode.succeed Authored
+
+        "shared" ->
+            Decode.succeed Shared
+
+        "curated" ->
+            Decode.succeed Curated
+
+        _ ->
+            Decode.fail (audienceType ++ " is not a recognized audienceType for Audience")
+
+
+
+-- MESSAGES
+
+
+type Msg
+    = SelectRootDirectory AudienceFolder
+    | SelectBreadcrumbItem AudienceFolder
+    | SelectSubDirectory AudienceFolder
+    | NoOp
+
+
+
+-- UPDATE
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    let
+        _ =
+            Debug.log "UPDATE msg: " msg
+
+        _ =
+            Debug.log "UPDATE model: " model
+    in
+        case msg of
+            SelectRootDirectory folder ->
+                ( { model
+                    | currentPath = [ folder ]
+                    , selectedFolder = Just folder
+                  }
+                , Cmd.none
+                )
+
+            SelectBreadcrumbItem folder ->
+                ( { model | selectedFolder = Just folder }, Cmd.none )
+
+            SelectSubDirectory folder ->
+                let
+                    newCurrentPath =
+                        setCurrentPath folder model.currentPath
+                in
+                    ( { model
+                        | currentPath = newCurrentPath
+                        , selectedFolder = Just folder
+                      }
+                    , Cmd.none
+                    )
```

Comment:
I'm still not sure if this is how it should work. I think one action/Msg should be enough just by looking at parameters.